pallets/pallet-asset-switch/src/lib.rs

```diff
@@ -838,6 +756,87 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		})
 	}
 
+	pub fn compute_xcm_for_switch(
+		inverted_universal_location: &Location,
+		from: &Location,
+		to: &Location,
+		amount: u128,
+		asset_id: &AssetId,
+		remote_asset_fee: &Asset,
+	) -> Xcm<()> {
+		let appendix = vec![
+			ReportHolding {
+				response_info: QueryResponseInfo {
+					destination: inverted_universal_location.clone(),
+					max_weight: Weight::zero(),
+					query_id: NextQueryId::<T, I>::get(),
+				},
+				// Include in the report the assets that were not transferred.
+				assets: AssetFilter::Wild(WildAsset::AllOf {
+					id: asset_id.clone(),
+					fun: WildFungibility::Fungible,
+				}),
+			},
+			// If the asset to transfer failed to transfer, re-put it back into our own account. Otherwise, if the
+			// transfer succeeded, this asset won't be present in the holding registry, hence this is effectively a
+			// no-op.
+			DepositAsset {
+				assets: AssetFilter::Wild(WildAsset::AllOf {
+					id: asset_id.clone(),
+					fun: WildFungibility::Fungible,
+				}),
+				beneficiary: inverted_universal_location.clone(),
+			},
+			RefundSurplus,
+			// Refund the XCM fee left to the user. We are purposefully only selecting the XCM fee asset, although
+			// there should be no cases in which any other assets are present in the holding registry.
+			DepositAsset {
+				assets: AssetFilter::Wild(WildAsset::AllOf {
+					id: remote_asset_fee.id.clone(),
+					fun: WildFungibility::Fungible,
+				}),
+				beneficiary: from.clone(),
+			},
+		]
+		.into();
+		// Steps performed:
+		// 1. Withdraw XCM fees from our SA
+		// 2. Buy execution
+		// 3. Set the appendix, executed regardless of the outcome of the transfer:
+		// 		3.1 Report back to our chain the assets in the holding registry. This will
+		// 			contain either only the XCM fee token in case of successful transfer, or the
+		//	 		XCM fee token + the amount of funds supposed to be transferred.
+		// 		3.2 Deposit the un-transferred asset (only if the transfer failed) back into
+		// our account.
+		//		3.3 Refund any surplus weight.
+		//		3.4 Deposit the remaining XCM fee assets in the user's account.
+		// 4. Withdraw the requested asset (this operation should be infallible since we
+		//    have full control of this balance)
+		// 5. Try to deposit the withdrawn asset into the user's account. This operation
+		//    could fail and the error is handled in the appendix.
```

Comment:
Should the docu maybe be moved as a docstring to the beginning of the function?

Reply:
https://github.com/KILTprotocol/kilt-node/pull/746/commits/53b5184c3f6c814334f53e104c5b894c640f0bc5